src/main/java/org/jahia/modules/sitemap/utils/Utils.java

```diff
@@ -96,95 +104,98 @@ public static String encodeSitemapLink(String URIPath, boolean shouldBeDecodedFi
-        List<String> excludedPath = new ArrayList<>();
-        JahiaUser guestUser = ServicesRegistry.getInstance().getJahiaUserManagerService().lookupUser(Constants.GUEST_USERNAME).getJahiaUser();
-        SitemapConfigService config = BundleUtils.getOsgiService(SitemapConfigService.class, null);
+    public static void generateSitemapEntries(Map<Locale, JCRSessionWrapper> sessionPerLocale, RenderContext renderContext, String rootPath, Set<String> sitemapRoots, Map<Locale, Set<SitemapEntry>> entriesByLocale, Map<String, Set<SitemapEntry>> entriesByPath) throws RepositoryException {
+         SitemapConfigService config = BundleUtils.getOsgiService(SitemapConfigService.class, null);
         if (config == null) {
             logger.error("Configuration service SitemapConfigService not revolved, check OSGi services status");
-            return Collections.EMPTY_SET;
+            return;
         }
-        JCRTemplate.getInstance().doExecute(guestUser, Constants.LIVE_WORKSPACE, locale, session -> {
-            // add root node into results
-            JCRNodeWrapper rootNode = session.getNode(rootPath);
-            if (isValidEntry(rootNode, renderContext)) {
-                result.add(buildSiteMapEntry(rootNode, locale, guestUser, renderContext));
-            }
-            // look for sub nodes
-            for (String nodeType : config.getIncludeContentTypes()) {
-                String queryFrom = String.format("select * FROM [%s] as sel WHERE ISDESCENDANTNODE(sel, '%s')", nodeType, StringUtils.replace(rootPath, "'", "''"));
-                new ScrollableQuery(500, session.getWorkspace().getQueryManager()
-                        .createQuery(queryFrom, Query.JCR_SQL2)).execute(
-                        new ScrollableQueryCallback<ScrollableQuery>() {
-                            @Override
-                            public boolean scroll() throws RepositoryException {
-                                for (NodeIterator iter = stepResult.getNodes(); iter.hasNext(); ) {
-                                    JCRNodeWrapper node = (JCRNodeWrapper) iter.nextNode();
-                                    if (node != null && node.isNodeType(DEDICATED_SITEMAP_MIXIN) && !node.getPath().equals(rootPath)) {
-                                        excludedPath.add(node.getPath());
-                                    } else if (node != null && isValidEntry(node, renderContext)) {
-                                        result.add(buildSiteMapEntry(node, locale, guestUser, renderContext));
-                                    }
+
+        JCRSessionWrapper session = sessionPerLocale.get(null);
+        // Get excluded path
+        String queryFromRootPath = String.format("select * FROM [%s] as sel WHERE ISDESCENDANTNODE(sel, '%s')", DEDICATED_SITEMAP_MIXIN, StringUtils.replace(rootPath, "'", "''"));
+        QueryResult results = session.getWorkspace().getQueryManager().createQuery(queryFromRootPath, Query.JCR_SQL2).execute();
```

Comment:
The results of this query seems to be never used

Reply:
thank you!